File: 프로그래머스/2/60057. 문자열 압축/문자열 압축.java
import java.lang.StringBuilder;
import java.util.ArrayList;
import java.util.List;

class Solution {

    private List<String> split(String s, int n) {
        List<String> result = new ArrayList<>();
        
        for (int start = 0; start <= s.length(); start += n) {
            int end = start + n;
            
            if (end > s.length()) {
                end = s.length();
            }
            
            result.add(s.substring(start, end));
        }
        
        return result;
    }
    
    private int compress(String s, int n) {
        StringBuilder sb = new StringBuilder();
        List<String> tokens = split(s, n);
        
        String prev = null;
        int count = 1;
        
        for (String token : tokens) {
            if (prev == null) {
                prev = token;
                continue;
            }
            
            if (token.equals(prev)) {
                count++;
                continue;
            }
            
            if (count > 1) sb.append(count);
            sb.append(prev);
            
            count = 1;
            prev = token;
        }
        
        if (count > 1) sb.append(count);
        sb.append(prev);
        
        return sb.length();
    }
    
    public int solution(String s) {
        if (s.length() == 1) {
            return 1;
        }
        
        int answer = s.length() + 1;
        
        for (int i = 1; i <= s.length() / 2; i++) {
            int length = compress(s, i);
            
            if (length < answer) {
                answer = length;
            }
        }
        
        return answer;
    }
}

File: 프로그래머스/2/60057. 문자열 압축/문자열 압축.py
from collections import deque


def solution(s):
    answer = len(s)
    
    for n in range(1, len(s) // 2 + 1):
        answer = min(answer, compress(s, n))
    
    return answer


def compress(s, n):
    q = deque()
    index = 0
    
    while index < len(s):
        q.append(s[index:index+n])
        index += n
    
    result = []
    prev = q.popleft()
    count = 1
    
    while q:
        cur = q.popleft()
        
        if prev == cur:
            count += 1
            continue
        
        result.append(prev if count == 1 else f'{count}{prev}')
        prev = cur
        count = 1

    result.append(prev if count == 1 else f'{count}{prev}')
    return len(''.join(result))
